passes/builders.nim

```diff
@@ -35,6 +35,18 @@ func add*[T](bu: var Builder[T], n: TreeNode[T]) =
   inc bu.buf[bu.parent].val
   bu.buf.add n
 
+func copyFrom*[T](bu: var Builder[T], tree: PackedTree[T], n: NodeIndex) =
+  ## Inserts the whole subtree from `tree` at `n` at the current buffer
+  ## position. `tree` has to use the same underlying storage for literal
+  ## data as the target tree.
+  # TODO: this procedure is misguided. The builder should not handle
+  #       things such as tree copies -- that's better left to ``ChangeSet``.
```

Comment:
Yep, that's exactly the reason. The `ChangeSet` application backend already somewhat supports this, but the tricky part is building an ergonomic and efficient interface around it. 

`ChangeSet` already has the `changeKind`, `insert`, `remove` procedures, and while they're good for one-off changes, they get cumbersome very quickly for some more sweeping and elaborate changes. They also cannot do things like, for example, turning `(A x y)` into `(B (C a x) (C b y))` (without creating temporary copies of `x` and `y`), which is akin to the changes `pass3` performs.

Reply:
I was thinking if `changeset` had a snapshot capability, it could internally materialize a tree and then keep applying changes atop that. I think that would allow stacking changes, but I'm not thinking too hard about efficiency, but focused on ergonomics.